src/ResourceCreator.ts

```diff
@@ -43,28 +43,28 @@ class ResourceCreator {
    * @param data                entity data from the Vuex store
    * @returns object            wrapped entity ready for use in a frontend component
    */
-  wrap (data: StoreData): ResourceInterface {
+  wrap<ResourceType extends ResourceInterface> (data: StoreData<ResourceType>): ResourceType {
     const meta = data._meta || { load: Promise.resolve(), loading: false }
 
     // Resource is loading --> return LoadingResource
     if (meta.loading) {
-      const loadResource = (meta.load as Promise<StoreData>).then(storeData => this.wrapData(storeData))
-      return new LoadingResource(loadResource, meta.self, this.config)
+      const loadResource = (meta.load as Promise<StoreData<ResourceType>>).then(storeData => this.wrapData<ResourceType>(storeData))
+      return new LoadingResource<ResourceType>(loadResource, meta.self, this.config) as unknown as ResourceType
 
     // Resource is not loading --> wrap actual data
     } else {
-      return this.wrapData(data)
+      return this.wrapData<ResourceType>(data)
     }
   }
 
-  wrapData (data: StoreData): ResourceInterface {
+  wrapData<ResourceType extends ResourceInterface<ResourceType>> (data: StoreData<ResourceType>): ResourceType {
     // Store data looks like a collection --> return CollectionInterface
-    if (isCollection(data)) {
-      return new Collection(data, this.apiActions, this, this.config) // these parameters are passed to Resource constructor
+    if (isCollection<ResourceType>(data)) {
+      return new Collection<ResourceType>(data as StoreDataCollection<ResourceType>, this.apiActions, this, this.config) as unknown as ResourceType// these parameters are passed to Resource constructor
 
     // else Store Data looks like an entity --> return normal Resource
     } else {
-      return new Resource(data, this.apiActions, this, this.config)
+      return new Resource<ResourceType>(data as StoreDataEntity<ResourceType>, this.apiActions, this, this.config) as unknown as ResourceType
```

Comment:
why can't you pass on `Resource<ResourceType>` which is the actual type of the return value?